numpy/core/src/multiarray/getset.c

```diff
@@ -492,58 +490,70 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
     }
 
 
-    if ((newtype->elsize != PyArray_DESCR(self)->elsize) &&
-            (PyArray_NDIM(self) == 0 ||
-             !PyArray_ISONESEGMENT(self) ||
-             PyDataType_HASSUBARRAY(newtype))) {
```

Comment:
There's a lot of unrelated failures in here that justify their own error message. `PyArray_ISONESEGMENT` becomes the checks for F and C contiguity further down.